common/scala/src/main/scala/org/apache/openwhisk/http/PoolingRestClient.scala

```diff
@@ -127,7 +134,13 @@ class PoolingRestClient(
       }
     }
 
-  def shutdown(): Future[Unit] = Future.unit
+  def shutdown(): Future[Unit] = {
+    killSwitch.shutdown()
+    Try(requestQueue.complete()).recover {
+      case t: IllegalStateException => logging.error(this, t.getMessage)
```

Comment:
An exception will be thrown if a stream is already completed, but there is no API to check the stream state.

Reply:
If the stream is already completed, I think this is a normal case.
Do we have to print an error log for this?

Reply:
For our service, we decided to have a log record of this exception as we can track such a situation and see if the number of such exceptions gets worse, which might help to investigate other issues. Also, personally, I don't like the approach of silently ignoring exceptions.

In our case, if the exception propagated to the dependable code it will break the k8s containers clean up which will lead to thousands of abandoned containers.

Reply:
In which exact case would it happen?
I thought it would always happen when we shut down the queue as we call `shtudown` of the kill switch.
So I thought changing the log level to info or warn would be enough for this.
But does this log stand for any abnormal situation?

Reply:
I'm not sure if `Killswitch` shuts it down all the time, in theory, it should, but I remember seeing an issue with memory leak only with having a kill switch and without `queue.complete`. By the number of these log messages and requests through the queue, it's hard to say if the kill switch does it all the time (maybe due to concurrent requests as well), but I saw a much smaller number of exceptions than requests that went through the queue.

I think it might also happen when some other place has already closed the queue.

Having just a warning message might be fine as well.